{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# About the data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "%matplotlib inline\n",
    "\n",
    "import os\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import random\n",
    "#import umap\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "from numpy.random import seed\n",
    "randomState = 123\n",
    "seed(randomState)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load arguments\n",
    "data_file = os.path.join(\n",
    "    os.path.dirname(os.getcwd()),\n",
    "    \"data\",\"pseudomonas\",\"train\", \"A\", \"all-pseudomonas-gene-normalized.zip\")\n",
    "\n",
    "metadata_file = os.path.join(\n",
    "    os.path.dirname(os.getcwd()),\n",
    "    \"metadata\",\"sample_annotations.tsv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1191, 5549)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Gene_symbol</th>\n",
       "      <th>PA0001</th>\n",
       "      <th>PA0002</th>\n",
       "      <th>PA0003</th>\n",
       "      <th>PA0004</th>\n",
       "      <th>PA0005</th>\n",
       "      <th>PA0006</th>\n",
       "      <th>PA0007</th>\n",
       "      <th>PA0008</th>\n",
       "      <th>PA0009</th>\n",
       "      <th>PA0010</th>\n",
       "      <th>...</th>\n",
       "      <th>PA5561</th>\n",
       "      <th>PA5562</th>\n",
       "      <th>PA5563</th>\n",
       "      <th>PA5564</th>\n",
       "      <th>PA5565</th>\n",
       "      <th>PA5566</th>\n",
       "      <th>PA5567</th>\n",
       "      <th>PA5568</th>\n",
       "      <th>PA5569</th>\n",
       "      <th>PA5570</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0.1_12hr_CSV86(Pae_G1a).CEL</th>\n",
       "      <td>0.472897</td>\n",
       "      <td>0.396658</td>\n",
       "      <td>0.253776</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.175640</td>\n",
       "      <td>0.554385</td>\n",
       "      <td>0.411370</td>\n",
       "      <td>0.382222</td>\n",
       "      <td>0.310144</td>\n",
       "      <td>0.642522</td>\n",
       "      <td>...</td>\n",
       "      <td>0.358597</td>\n",
       "      <td>0.390048</td>\n",
       "      <td>0.457406</td>\n",
       "      <td>0.684082</td>\n",
       "      <td>0.338351</td>\n",
       "      <td>0.608325</td>\n",
       "      <td>0.643496</td>\n",
       "      <td>0.276075</td>\n",
       "      <td>0.112773</td>\n",
       "      <td>0.145170</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0.1_2hr_CSV86(Pae_G1a).CEL</th>\n",
       "      <td>0.262346</td>\n",
       "      <td>0.086216</td>\n",
       "      <td>0.359853</td>\n",
       "      <td>0.439214</td>\n",
       "      <td>0.269749</td>\n",
       "      <td>0.768433</td>\n",
       "      <td>0.212505</td>\n",
       "      <td>0.062043</td>\n",
       "      <td>0.567695</td>\n",
       "      <td>0.467073</td>\n",
       "      <td>...</td>\n",
       "      <td>0.358504</td>\n",
       "      <td>0.414206</td>\n",
       "      <td>0.389879</td>\n",
       "      <td>0.477693</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.479385</td>\n",
       "      <td>0.154471</td>\n",
       "      <td>0.140891</td>\n",
       "      <td>0.167505</td>\n",
       "      <td>0.157060</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0.1_6hr_CSV86(Pae_G1a).CEL</th>\n",
       "      <td>0.473658</td>\n",
       "      <td>0.244862</td>\n",
       "      <td>0.330750</td>\n",
       "      <td>0.097697</td>\n",
       "      <td>0.387226</td>\n",
       "      <td>0.328319</td>\n",
       "      <td>0.228820</td>\n",
       "      <td>0.330039</td>\n",
       "      <td>0.318081</td>\n",
       "      <td>0.512864</td>\n",
       "      <td>...</td>\n",
       "      <td>0.180744</td>\n",
       "      <td>0.380741</td>\n",
       "      <td>0.173501</td>\n",
       "      <td>0.251571</td>\n",
       "      <td>0.182793</td>\n",
       "      <td>0.528301</td>\n",
       "      <td>0.504985</td>\n",
       "      <td>0.499782</td>\n",
       "      <td>0.061106</td>\n",
       "      <td>0.365612</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0.1_7hr_CSV86(Pae_G1a).CEL</th>\n",
       "      <td>0.439273</td>\n",
       "      <td>0.343402</td>\n",
       "      <td>0.192698</td>\n",
       "      <td>0.274677</td>\n",
       "      <td>0.628979</td>\n",
       "      <td>0.553796</td>\n",
       "      <td>0.431391</td>\n",
       "      <td>0.363480</td>\n",
       "      <td>0.385721</td>\n",
       "      <td>0.094584</td>\n",
       "      <td>...</td>\n",
       "      <td>0.346837</td>\n",
       "      <td>0.153927</td>\n",
       "      <td>0.067349</td>\n",
       "      <td>0.319723</td>\n",
       "      <td>0.282442</td>\n",
       "      <td>0.490655</td>\n",
       "      <td>0.531415</td>\n",
       "      <td>0.153880</td>\n",
       "      <td>0.132333</td>\n",
       "      <td>0.260087</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0.1_9hr_CSV86(Pae_G1a).CEL</th>\n",
       "      <td>0.220827</td>\n",
       "      <td>0.145525</td>\n",
       "      <td>0.437803</td>\n",
       "      <td>0.293201</td>\n",
       "      <td>0.635120</td>\n",
       "      <td>0.462893</td>\n",
       "      <td>0.488733</td>\n",
       "      <td>0.309584</td>\n",
       "      <td>0.318646</td>\n",
       "      <td>0.591914</td>\n",
       "      <td>...</td>\n",
       "      <td>0.237726</td>\n",
       "      <td>0.301945</td>\n",
       "      <td>0.070222</td>\n",
       "      <td>0.513605</td>\n",
       "      <td>0.114277</td>\n",
       "      <td>0.360259</td>\n",
       "      <td>0.386868</td>\n",
       "      <td>0.223995</td>\n",
       "      <td>0.105343</td>\n",
       "      <td>0.102088</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 5549 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Gene_symbol                    PA0001    PA0002    PA0003    PA0004    PA0005  \\\n",
       "0.1_12hr_CSV86(Pae_G1a).CEL  0.472897  0.396658  0.253776  0.000000  0.175640   \n",
       "0.1_2hr_CSV86(Pae_G1a).CEL   0.262346  0.086216  0.359853  0.439214  0.269749   \n",
       "0.1_6hr_CSV86(Pae_G1a).CEL   0.473658  0.244862  0.330750  0.097697  0.387226   \n",
       "0.1_7hr_CSV86(Pae_G1a).CEL   0.439273  0.343402  0.192698  0.274677  0.628979   \n",
       "0.1_9hr_CSV86(Pae_G1a).CEL   0.220827  0.145525  0.437803  0.293201  0.635120   \n",
       "\n",
       "Gene_symbol                    PA0006    PA0007    PA0008    PA0009    PA0010  \\\n",
       "0.1_12hr_CSV86(Pae_G1a).CEL  0.554385  0.411370  0.382222  0.310144  0.642522   \n",
       "0.1_2hr_CSV86(Pae_G1a).CEL   0.768433  0.212505  0.062043  0.567695  0.467073   \n",
       "0.1_6hr_CSV86(Pae_G1a).CEL   0.328319  0.228820  0.330039  0.318081  0.512864   \n",
       "0.1_7hr_CSV86(Pae_G1a).CEL   0.553796  0.431391  0.363480  0.385721  0.094584   \n",
       "0.1_9hr_CSV86(Pae_G1a).CEL   0.462893  0.488733  0.309584  0.318646  0.591914   \n",
       "\n",
       "Gene_symbol                    ...       PA5561    PA5562    PA5563    PA5564  \\\n",
       "0.1_12hr_CSV86(Pae_G1a).CEL    ...     0.358597  0.390048  0.457406  0.684082   \n",
       "0.1_2hr_CSV86(Pae_G1a).CEL     ...     0.358504  0.414206  0.389879  0.477693   \n",
       "0.1_6hr_CSV86(Pae_G1a).CEL     ...     0.180744  0.380741  0.173501  0.251571   \n",
       "0.1_7hr_CSV86(Pae_G1a).CEL     ...     0.346837  0.153927  0.067349  0.319723   \n",
       "0.1_9hr_CSV86(Pae_G1a).CEL     ...     0.237726  0.301945  0.070222  0.513605   \n",
       "\n",
       "Gene_symbol                    PA5565    PA5566    PA5567    PA5568    PA5569  \\\n",
       "0.1_12hr_CSV86(Pae_G1a).CEL  0.338351  0.608325  0.643496  0.276075  0.112773   \n",
       "0.1_2hr_CSV86(Pae_G1a).CEL   0.000000  0.479385  0.154471  0.140891  0.167505   \n",
       "0.1_6hr_CSV86(Pae_G1a).CEL   0.182793  0.528301  0.504985  0.499782  0.061106   \n",
       "0.1_7hr_CSV86(Pae_G1a).CEL   0.282442  0.490655  0.531415  0.153880  0.132333   \n",
       "0.1_9hr_CSV86(Pae_G1a).CEL   0.114277  0.360259  0.386868  0.223995  0.105343   \n",
       "\n",
       "Gene_symbol                    PA5570  \n",
       "0.1_12hr_CSV86(Pae_G1a).CEL  0.145170  \n",
       "0.1_2hr_CSV86(Pae_G1a).CEL   0.157060  \n",
       "0.1_6hr_CSV86(Pae_G1a).CEL   0.365612  \n",
       "0.1_7hr_CSV86(Pae_G1a).CEL   0.260087  \n",
       "0.1_9hr_CSV86(Pae_G1a).CEL   0.102088  \n",
       "\n",
       "[5 rows x 5549 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Read in data\n",
    "data = pd.read_table(data_file, header=0, sep='\\t', index_col=0, compression='zip').T\n",
    "original_shape = data.shape\n",
    "print(original_shape)\n",
    "data.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1217, 17)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_name</th>\n",
       "      <th>ml_data_source</th>\n",
       "      <th>description</th>\n",
       "      <th>nucleic_acid</th>\n",
       "      <th>medium</th>\n",
       "      <th>genotype</th>\n",
       "      <th>od</th>\n",
       "      <th>growth_setting_1</th>\n",
       "      <th>growth_setting_2</th>\n",
       "      <th>strain</th>\n",
       "      <th>temperature</th>\n",
       "      <th>treatment</th>\n",
       "      <th>additional_notes</th>\n",
       "      <th>variant_phenotype</th>\n",
       "      <th>abx_marker</th>\n",
       "      <th>biotic_int_lv_2</th>\n",
       "      <th>biotic_int_lv_1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>experiment</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>E-GEOD-46947</th>\n",
       "      <td>GSM1141730 1</td>\n",
       "      <td>GSM1141730_PA01_ZnO_PZO_.CEL</td>\n",
       "      <td>Pseudomonas aeruginosa PAO1 LB aerated  5 h wi...</td>\n",
       "      <td>RNA</td>\n",
       "      <td>LB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>planktonic</td>\n",
       "      <td>aerated</td>\n",
       "      <td>PAO1</td>\n",
       "      <td>37.0</td>\n",
       "      <td>1 mM ZnO nanoparticles</td>\n",
       "      <td>Grown for 5h</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>E-GEOD-46947</th>\n",
       "      <td>GSM1141729 1</td>\n",
       "      <td>GSM1141729_PA01_none_PC_.CEL</td>\n",
       "      <td>Pseudomonas aeruginosa PAO1 LB aerated  5 h</td>\n",
       "      <td>RNA</td>\n",
       "      <td>LB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>planktonic</td>\n",
       "      <td>aerated</td>\n",
       "      <td>PAO1</td>\n",
       "      <td>37.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Grown for 5h</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>E-GEOD-65882</th>\n",
       "      <td>GSM1608059 1</td>\n",
       "      <td>GSM1608059_Planktonic_1.CEL</td>\n",
       "      <td>PAO1 WT. Planktonic. Rep1</td>\n",
       "      <td>RNA</td>\n",
       "      <td>PBM plus 1 g / L glucose.</td>\n",
       "      <td>WT</td>\n",
       "      <td>0.26</td>\n",
       "      <td>Planktonic</td>\n",
       "      <td>Aerated</td>\n",
       "      <td>PAO1</td>\n",
       "      <td>37.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Grown shaking at 200rpm</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>E-GEOD-65882</th>\n",
       "      <td>GSM1608060 1</td>\n",
       "      <td>GSM1608060_Planktonic_2.CEL</td>\n",
       "      <td>PAO1 WT. Planktonic. Rep2</td>\n",
       "      <td>RNA</td>\n",
       "      <td>PBM plus 1 g / L glucose.</td>\n",
       "      <td>WT</td>\n",
       "      <td>0.26</td>\n",
       "      <td>Planktonic</td>\n",
       "      <td>Aerated</td>\n",
       "      <td>PAO1</td>\n",
       "      <td>37.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Grown shaking at 200rpm</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>E-GEOD-65882</th>\n",
       "      <td>GSM1608061 1</td>\n",
       "      <td>GSM1608061_Planktonic_3.CEL</td>\n",
       "      <td>PAO1 WT. Planktonic. Rep3</td>\n",
       "      <td>RNA</td>\n",
       "      <td>PBM plus 1 g / L glucose.</td>\n",
       "      <td>WT</td>\n",
       "      <td>0.26</td>\n",
       "      <td>Planktonic</td>\n",
       "      <td>Aerated</td>\n",
       "      <td>PAO1</td>\n",
       "      <td>37.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Grown shaking at 200rpm</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               sample_name                ml_data_source  \\\n",
       "experiment                                                 \n",
       "E-GEOD-46947  GSM1141730 1  GSM1141730_PA01_ZnO_PZO_.CEL   \n",
       "E-GEOD-46947  GSM1141729 1  GSM1141729_PA01_none_PC_.CEL   \n",
       "E-GEOD-65882  GSM1608059 1   GSM1608059_Planktonic_1.CEL   \n",
       "E-GEOD-65882  GSM1608060 1   GSM1608060_Planktonic_2.CEL   \n",
       "E-GEOD-65882  GSM1608061 1   GSM1608061_Planktonic_3.CEL   \n",
       "\n",
       "                                                    description nucleic_acid  \\\n",
       "experiment                                                                     \n",
       "E-GEOD-46947  Pseudomonas aeruginosa PAO1 LB aerated  5 h wi...          RNA   \n",
       "E-GEOD-46947        Pseudomonas aeruginosa PAO1 LB aerated  5 h          RNA   \n",
       "E-GEOD-65882                          PAO1 WT. Planktonic. Rep1          RNA   \n",
       "E-GEOD-65882                          PAO1 WT. Planktonic. Rep2          RNA   \n",
       "E-GEOD-65882                          PAO1 WT. Planktonic. Rep3          RNA   \n",
       "\n",
       "                                  medium genotype    od growth_setting_1  \\\n",
       "experiment                                                                 \n",
       "E-GEOD-46947                          LB      NaN   NaN       planktonic   \n",
       "E-GEOD-46947                          LB      NaN   NaN       planktonic   \n",
       "E-GEOD-65882  PBM plus 1 g / L glucose.        WT  0.26      Planktonic    \n",
       "E-GEOD-65882  PBM plus 1 g / L glucose.        WT  0.26      Planktonic    \n",
       "E-GEOD-65882  PBM plus 1 g / L glucose.        WT  0.26      Planktonic    \n",
       "\n",
       "             growth_setting_2 strain  temperature               treatment  \\\n",
       "experiment                                                                  \n",
       "E-GEOD-46947          aerated   PAO1         37.0  1 mM ZnO nanoparticles   \n",
       "E-GEOD-46947          aerated   PAO1         37.0                     NaN   \n",
       "E-GEOD-65882          Aerated   PAO1         37.0                     NaN   \n",
       "E-GEOD-65882          Aerated   PAO1         37.0                     NaN   \n",
       "E-GEOD-65882          Aerated   PAO1         37.0                     NaN   \n",
       "\n",
       "                     additional_notes variant_phenotype abx_marker  \\\n",
       "experiment                                                           \n",
       "E-GEOD-46947             Grown for 5h               NaN        NaN   \n",
       "E-GEOD-46947             Grown for 5h               NaN        NaN   \n",
       "E-GEOD-65882  Grown shaking at 200rpm               NaN        NaN   \n",
       "E-GEOD-65882  Grown shaking at 200rpm               NaN        NaN   \n",
       "E-GEOD-65882  Grown shaking at 200rpm               NaN        NaN   \n",
       "\n",
       "             biotic_int_lv_2 biotic_int_lv_1  \n",
       "experiment                                    \n",
       "E-GEOD-46947             NaN             NaN  \n",
       "E-GEOD-46947             NaN             NaN  \n",
       "E-GEOD-65882             NaN             NaN  \n",
       "E-GEOD-65882             NaN             NaN  \n",
       "E-GEOD-65882             NaN             NaN  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Read in metadata\n",
    "metadata = pd.read_table(metadata_file, header=0, sep='\\t', index_col=0)\n",
    "metadata_shape = metadata.shape\n",
    "print(metadata_shape)\n",
    "metadata.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## What does the clustering look like between planktonic vs biofilm in Gene space\n",
    "\n",
    "How much heterogeneity exists between samples?\n",
    "How many genes contribute to the difference?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Label samples by planktonic vs biofilm\n",
    "\n",
    "# Since our simulation set all genes in set A to be the same value for a give sample\n",
    "# we can consider a single gene in set A to query by\n",
    "rep_gene_A = geneSetA.iloc[0][0]\n",
    "geneA_exp = sim_data[rep_gene_A]\n",
    "\n",
    "sample_id = sim_data.index\n",
    "\n",
    "# Bin gene A expression\n",
    "geneA_exp_labeled = sim_data.assign(\n",
    "    rep_geneA=(\n",
    "        list( \n",
    "            map(\n",
    "                lambda x:\n",
    "                '1' if 0 < x and x <=0.1 \n",
    "                else '2' if 0.1< x and x <=0.2 \n",
    "                else '3' if 0.2<x and x<=0.3\n",
    "                else '4' if 0.3<x  and x<=0.4\n",
    "                else '5' if 0.4<x and x<=0.5\n",
    "                else '6' if 0.5<x and x<=0.6\n",
    "                else '7' if 0.6<x and x<=0.7\n",
    "                else '8' if 0.7<x and x<=0.8\n",
    "                else '9' if 0.8<x and x<=0.9\n",
    "                else '10',\n",
    "                geneA_exp\n",
    "            )      \n",
    "        )\n",
    "    )\n",
    ")\n",
    "geneA_exp_labeled = geneA_exp_labeled.astype({\"rep_geneA\": int})\n",
    "geneA_exp_labeled.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Plot gene expression in gene space\n",
    "Each dot is a sample. Each sample is colored based on its expression of gene A\n",
    "\n",
    "In the legend 1 ~ gene A expression is (0.0, 0.1], 2 ~ gene A expression is (0.1, 0.2], etc."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# UMAP embedding of raw gene space data\n",
    "embedding = umap.UMAP().fit_transform(sim_data)\n",
    "embedding.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# UMAP plot of raw gene expression data\n",
    "geneA_exp_labeled = geneA_exp_labeled.assign(sample_index=list(range(geneA_exp_labeled.shape[0])))\n",
    "for x in geneA_exp_labeled.rep_geneA.sort_values().unique():\n",
    "    plt.scatter(\n",
    "        embedding[geneA_exp_labeled.query(\"rep_geneA == @x\").sample_index.values, 0], \n",
    "        embedding[geneA_exp_labeled.query(\"rep_geneA == @x\").sample_index.values, 1], \n",
    "        c=sns.color_palette()[x-1],\n",
    "        alpha=0.7,\n",
    "        label=str(x)\n",
    "    )\n",
    "plt.gca().set_aspect('equal', 'datalim')\n",
    "plt.title('UMAP projection of gene expression data in GENE space', fontsize=14)\n",
    "plt.legend()\n",
    "plt.savefig(geneSpace_file, dpi=300)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
